myia/grad.py

```diff
@@ -0,0 +1,284 @@
+            return rval
+        else:
+            return None
+
+    def rho(self, node, graph):
+        """Compute expression for gradient wrt node and graph."""
+
+        # We index with node and graph because the same node may have multiple
+        # sensitivity variables, one for each graph that refers to the original.
+        key = (node, graph)
+        if key in self.sensitivity_nodes:
+            return self.sensitivity_nodes[key]
+
+        bg = self.backpropagator_graphs[graph]
+
+        # We will accumulate all gradient contributions here.
+        contribs = []
+
+        for user, idx in node.uses:
+            # Each use of a variable contributes to its gradient.
+            ug = user.graph
+            if ug is graph:
+                # A use in the same graph: we get the backpropagator expression
+                # and we use the argument index to extract the right
+                # contribution.
+                contrib = Apply([index,
+                                self.bprop_step(user),
+                                Constant(idx)], bg)
+                contribs.append(contrib)
+            elif self.nest.nested_in(ug, graph):
+                # A use in a different graph. This contribution will come from
+                # the sensitivity to the closure for that graph.
+                for graph_ct in self.graph_to_ct[ug]:
+                    fvs = self.fv_order[ug]
+                    assert node in fvs
+                    # This is the index of this node in the graph's free
+                    # variables.
+                    idx = fvs.index(node)
+                    # We get the sensitivity wrt the closure using `rho` on the
+                    # constant and this graph. Concretely this means we will
+                    # look at the uses of the closure in this graph. Or if the
+                    # closure is returned, this will be the sensitivity wrt the
+                    # output. We index this sensitivity with idx to get the
+                    # contribution we seek.
+                    contrib = Apply([index,
+                                    self.rho(graph_ct, graph),
+                                    Constant(idx)], bg)
+                    contribs.append(contrib)
+            else:
+                pass
+
+        # NOTE: The order of nodes in contribs is not deterministic, because
+        # the order of users isn't. In theory that doesn't matter, because we
+        # add them all, but in practice there could be slight numerical
+        # differences.
+
+        if len(contribs) == 0:
+            # No contributions means a gradient of zero, naturally.
+            sens = Constant(0)  # TODO: should be zeros_like(node)
+        elif len(contribs) == 1:
+            # A single contribution can be used directly.
+            sens, = contribs
+        else:
+            # More than one contribution must be added trogether.
+            sens = Apply([add, *contribs], bg)
```

Comment:
Just a note that although `add` is currently implemented as `sum` in the VM, it's still an `add`, and we probably should stick to these simpler semantics for now and leave `sum` for a later part in the optimization pipeline.

Reply:
Sure.